portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -30,21 +34,49 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  // Hard coded comment data for now.
-  private final List<Comment> comments =
-      Collections.unmodifiableList(
-          Arrays.asList(
-              new Comment("Fred", new Date(), "This is comment number 1"),
-              new Comment("George", new Date(), "This is comment number 2"),
-              new Comment("Ron", new Date(), "This is comment number 3")));
-
-  private final String CONTENT_TYPE = "text/json;";
+  private final String COMMENT_ENTITY_NAME = "Comment";
+  private final ArrayList<Comment> comments = new ArrayList<Comment>();
+  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  private final String GET_CONTENT_TYPE = "text/json;";
   private final Gson gson = new Gson();
+  private final String POST_MESSAGE_PARAMETER = "body";
+  private final String POST_NAME_PARAMETER = "name";
+  private final String POST_REDIRECT_URL = "/";
+  private final String PROPERTY_COMMENT_CREATED_NAME = "created";
+  private final String PROPERTY_COMMENT_MESSAGE_NAME = "message";
+  private final String PROPERTY_COMMENT_NAME = "name";
+  private final Query query =
+      new Query(this.COMMENT_ENTITY_NAME)
+          .addSort(this.PROPERTY_COMMENT_CREATED_NAME, SortDirection.DESCENDING);
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    this.comments.clear();
+    PreparedQuery results = this.datastore.prepare(this.query);
+    for (Entity comment : results.asIterable()) {
+      final String name = (String) comment.getProperty(this.PROPERTY_COMMENT_NAME);
+      final Date created = new Date((long) comment.getProperty(this.PROPERTY_COMMENT_CREATED_NAME));
+      final String message = (String) comment.getProperty(this.PROPERTY_COMMENT_MESSAGE_NAME);
+      this.comments.add(new Comment(name, created, message));
+    }
     String output = gson.toJson(this.comments);
-    response.setContentType(this.CONTENT_TYPE);
+    response.setContentType(this.GET_CONTENT_TYPE);
     response.getWriter().println(output);
   }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    final String name = request.getParameter(this.POST_NAME_PARAMETER);
+    final String body = request.getParameter(POST_MESSAGE_PARAMETER);
+    if (name == null || body == null) { // Someone sending a bad form.
+      return;
+    }
```

Comment:
Set response code and error message.

Reply:
I merged master into this branch because it already has a commit that sets the response code and error message.